src/Interpreters/JoinExpressionActions.cpp

```diff
@@ -0,0 +1,514 @@
+        case JoinConditionOperator::GreaterOrEquals: return ">=";
+        case JoinConditionOperator::And: return "AND";
+        case JoinConditionOperator::Or: return "OR";
+        case JoinConditionOperator::Unknown: break;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Illegal value for JoinConditionOperator: {}", static_cast<Int32>(op));
+}
+
+
+String toString(const BitSet & bitset)
+{
+    std::string str;
+    boost::to_string(bitset.bitset, str);
+    return str;
+}
+
+int BitSet::findFirstSet() const
+{
+    if (bitset.none())
+        return -1;
+
+    size_t pos = bitset.find_first();
+    size_t maxPos = pos;
+
+    while (pos != Base::npos)
+    {
+        maxPos = pos;
+        pos = bitset.find_next(pos);
+    }
+
+    if (maxPos > std::numeric_limits<int>::max())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "BitSet is too large");
+
+    return static_cast<int>(maxPos);
+}
+
+size_t BitSet::hashImpl() const
+{
+    if (likely(bitset.size() <= std::numeric_limits<unsigned long>::digits))
+        /// false positive clang-tidy warning for code in boost::dynamic_bitset
+        return std::hash<uint64_t>()(bitset.to_ulong()); /// NOLINT
+
+
+    UInt64 hash = 0;
+    auto pos = bitset.find_first();
+    while (pos != Base::npos)
+    {
+        intHashCRC32(hash, pos);
+        pos = bitset.find_next(pos);
+    }
+    return hash;
+}
+
+struct JoinExpressionActions::Data : boost::noncopyable
+{
+    using NodeToSourceMapping = std::unordered_map<NodeRawPtr, BitSet>;
+
+    Data(ActionsDAG && actions_dag_, NodeToSourceMapping && expression_sources_)
+        : actions_dag(std::move(actions_dag_)), expression_sources(std::move(expression_sources_))
+    {
+    }
+
+    BitSet lhs_rels;
+    BitSet rhs_rels;
+    ActionsDAG actions_dag;
+    NodeToSourceMapping expression_sources;
+};
+
+JoinExpressionActions::JoinExpressionActions()
+    : data(std::make_shared<Data>(ActionsDAG(), Data::NodeToSourceMapping()))
+{
+}
+
+JoinExpressionActions::JoinExpressionActions(const Block & left_header, const Block & right_header)
+{
+    Data::NodeToSourceMapping expression_sources;
+    ActionsDAG actions_dag;
+
+    for (const auto & column : left_header)
+    {
+        const auto * node = &actions_dag.addInput(column.name, column.type);
+        expression_sources[node].set(0);
+    }
+
+    for (const auto & column : right_header)
+    {
+        const auto * node = &actions_dag.addInput(column.name, column.type);
+        expression_sources[node].set(1);
+    }
+
+    data = std::make_shared<Data>(std::move(actions_dag), std::move(expression_sources));
+}
+
+JoinExpressionActions::JoinExpressionActions(const Block & left_header, const Block & right_header, ActionsDAG && actions_dag_)
+{
+    Data::NodeToSourceMapping expression_sources;
+
+    const auto & input_nodes = actions_dag_.getInputs();
+    if (input_nodes.size() != left_header.columns() + right_header.columns())
```

Comment:
Todo: Here, we can (probably) enforce an even stricter invariant that inputs are ordered in the same way.
(or just use size_t to split left and right inputs)